queasars/minimum_eigensolvers/base/termination_criteria.py

```diff
@@ -109,3 +110,184 @@ def check_termination(
+        return False
+
+
+def _median_hausdorff_distance_by_expectation_value(
+    result_1: BasePopulationEvaluationResult, result_2: BasePopulationEvaluationResult
+) -> float:
+    """
+    Calculates the median Hausdorff distance between the expectation values in two population evaluation results
+
+    :arg result_1: first evaluation result to calculate the distance on
+    :type result_1: BasePopulationEvaluationResult
+    :arg result_2: second evaluation result to calculate the distance on
+    :type result_2: BasePopulationEvaluationResult
+    :return: the median Hausdorff distance between the expectation values in two population evaluation results
+    """
+
+    def distance(from_expectations: list[float], to_expectations: list[float]) -> float:
+        distances: list[float] = []
+        for from_expectation in from_expectations:
+            distances.append(min(abs(from_expectation - to_expectation) for to_expectation in to_expectations))
+        return cast(float, median(distances))
+
+    expectations_1: list[float] = [exp for exp in result_1.expectation_values if exp is not None]
+    expectations_2: list[float] = [exp for exp in result_2.expectation_values if exp is not None]
+    return max(distance(expectations_1, expectations_2), distance(expectations_2, expectations_1))
+
+
+class PopulationChangeTolerance(EvolvingAnsatzMinimumEigensolverBaseTerminationCriterion):
+    """
+    Termination criterion which terminates, if the current generation has not changed sufficiently from
+    the last generation. This change is taken as the maximum of the absolute change of the expectation value of
+    the best individual from the last and current generation and the median Hausdorff distance between the
+    expectation values of the last and current generation. This means this criterion terminates if the best individual
+    AND the population as a whole do not change sufficiently. This termination criterion can be configured to only
+    terminate if the change falls below the threshold value multiple times consecutively
+
+    :param minimum_change: threshold value for the absolute change in expectation value, below which this criterion
+        chooses to terminate the optimization
+    :type minimum_change: float
+    :param allowed_consecutive_violations: determines how often the threshold value can be violated consecutively
+        before this criterion chooses to terminate. If set to 0, this terminates the first time the change falls
+        below the threshold value. If set to 2 for example, this terminates the first time the change falls below
+        the threshold three consecutive times. Must be at least 0.
+    :type allowed_consecutive_violations: int
+    """
+
+    def __init__(self, minimum_change: float, allowed_consecutive_violations: int):
+        super().__init__()
+        self._minimum_change: float = minimum_change
+        self._allowed_consecutive_violations: int = allowed_consecutive_violations
```

Comment:
Maybe verify that `allowed_consecutive_violations >= 0`?

Reply:
Done.